### PR TITLE
gRPC: Add operation to stream schema changes

### PR DESCRIPTION
**Pull request by olim7t**
**What this PR does**:
Add gRPC operation to stream schema changes.

**Which issue(s) this PR fixes**:
N/A

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated

### CODE REVIEW

File: grpc-proto/proto/stargate.proto
Comment (olim7t) [inline grpc-proto/proto/stargate.proto:35]: We might want to use a custom message here. Even if it's still empty now, that leaves us a way to add parameters in the future if needed.

File: testing/src/main/java/io/stargate/it/grpc/SchemaChangesTest.java
Comment (olim7t) [inline testing/src/main/java/io/stargate/it/grpc/SchemaChangesTest.java:49]: This is weird. I've tried with UDTs as well, and same thing, CREATE does not generate a notification but DROP does.

**Review by olim7t — COMMENTED** [review]
Work in progress, I'm still experimenting with the idea.

The server never completes the request, it's up to the client to cancel it when it's done.

File: testing/src/main/java/io/stargate/it/grpc/SchemaChangesTest.java
Comment (ivansenic) [reply testing/src/main/java/io/stargate/it/grpc/SchemaChangesTest.java:49]: because the listener is not yet registered once you do the `execute`, you call the async api with the GRPC and return thus immediately, it's not guaranteed that the execute will be done after the `persistence.registerEventListener(this);` is called in the `SchemaChangesHandler`

**Comment by jeffreyscarpenter** [comment]
Should we target this for v2.0.0 branch instead of master?

File: testing/src/main/java/io/stargate/it/grpc/SchemaChangesTest.java
Comment (olim7t) [reply testing/src/main/java/io/stargate/it/grpc/SchemaChangesTest.java:49]: Good catch! Pausing after line 46 does indeed fix the issue.

**Comment by olim7t** [comment]
> Should we target this for v2.0.0 branch instead of master?

It can't go on either side. But since we've been merging master->v2.0.0 so far, I think it's better to target master for things that are going to be on both sides.

**Comment by jeffreyscarpenter** [comment]
> > Should we target this for v2.0.0 branch instead of master?
> 
> It can't go on either side. But since we've been merging master->v2.0.0 so far, I think it's better to target master for things that are going to be on both sides.

I thought we had not decided for sure on adding schema operations to public gRPC yet, which is why we were doing this schema work in v2.0.0 first.

File: grpc-proto/proto/stargate.proto
Comment (olim7t) [reply grpc-proto/proto/stargate.proto]: Done.

File: grpc/src/main/java/io/stargate/grpc/service/GrpcService.java
Comment (olim7t) [reply grpc/src/main/java/io/stargate/grpc/service/GrpcService.java:78]: I'm going to take a closer look at this.
Event listeners are registered with the Persistence directly, there's no need to open a connection.

If tenants is what you have in mind, I'm not sure if we need to filter here. Event handling is generally a VM-wide concern, so it spans multiple tenant contexts. Therefore what we need is an "internal" view that notifies for all tenants. At least that's how it works in the GraphQL cache, our only existing use case so far.

File: grpc/src/main/java/io/stargate/grpc/service/interceptors/NewConnectionInterceptor.java
Comment (olim7t) [inline grpc/src/main/java/io/stargate/grpc/service/interceptors/NewConnectionInterceptor.java:92]: This is not super elegant...
Maybe creating the connection doesn't matter that much, since connections are cached and the same client is likely to invoke other methods than this one.

File: grpc/src/main/java/io/stargate/grpc/service/interceptors/NewConnectionInterceptor.java
Comment (tomekl007) [inline grpc/src/main/java/io/stargate/grpc/service/interceptors/NewConnectionInterceptor.java:88]: why do we need this special condition?

Comment (olim7t) [reply grpc/src/main/java/io/stargate/grpc/service/interceptors/NewConnectionInterceptor.java:88]: That operation interacts with the `Persistence` directly, it doesn't need the `Connnection`. So the idea was to avoid creating it.
But maybe that's overkill (see my other comment).

**Review by jeffreyscarpenter — APPROVED** [review]

File: sgv2-service-common/src/main/java/io/stargate/sgv2/common/grpc/InitReplayingStreamObserver.java
Comment (olim7t) [inline sgv2-service-common/src/main/java/io/stargate/sgv2/common/grpc/InitReplayingStreamObserver.java:32]: This is complicated but I can't think of a simpler solution.

I suspect calls to `onNext()` are always happening from the same Netty event loop, but `markReady()` definitely isn't, so we need proper synchronization.

File: sgv2-service-common/src/main/java/io/stargate/sgv2/common/schema/SchemaCache.java
Comment (olim7t) [inline sgv2-service-common/src/main/java/io/stargate/sgv2/common/schema/SchemaCache.java:167]: Unlike `init()`, we don't want to block here, we're on a gRPC thread.

File: sgv2-service-common/src/main/java/io/stargate/sgv2/common/schema/SchemaCache.java
Comment (olim7t) [inline sgv2-service-common/src/main/java/io/stargate/sgv2/common/schema/SchemaCache.java:174]: I'm not sure what else to do here. Maybe some kind of periodic retry, but we don't have access to an executor.

File: grpc/src/main/java/io/stargate/grpc/service/GrpcService.java
Comment (tatu-at-datastax) [reply grpc/src/main/java/io/stargate/grpc/service/GrpcService.java]: We probably cannot really afford listener-per-tenant so it should be global one.

One possible solution is that update would only be used to invalidate entry (regardless of tenant), but lookups would be on behalf of tenant when needed (not part of notification handling).

**Comment by tatu-at-datastax** [comment]
I think I'd like to get this merged. The only (?) question I have is wrt name handling of keyspaces, tenant ids, but that probably needs to be discussed outside this PR.

**Comment by olim7t** [comment]
> The only (?) question I have is wrt name handling of keyspaces, tenant ids, but that probably needs to be discussed outside this PR.

Agreed. The current implementation will return "decorated" keyspace names in a multi-tenant environment, we'll need to decide if we process them here or leave that to the caller.

**Comment by tatu-at-datastax** [comment]
> > The only (?) question I have is wrt name handling of keyspaces, tenant ids, but that probably needs to be discussed outside this PR.
> 
> Agreed. The current implementation will return "decorated" keyspace names in a multi-tenant environment, we'll need to decide if we process them here or leave that to the caller.

I like the idea of scoped identifier where unmodified `name` is returned along with string `tenant id`.
Probably best done with follow-up work so it does not block merging, and since we don't quite know how this would interact with multi-tenant backends.

**Pull request merged**
